rfcs/0049-flakes.md

```diff
@@ -0,0 +1,500 @@
+depends on the Nixpkgs flake and provides a package (i.e. an
+installable derivation) and a NixOS module.
+
+```
+{
+  description = "A filesystem that fetches DWARF debug info from the Internet on demand";
+
+  outputs = { self, nixpkgs }: rec {
+    packages.x86_64-linux.dwarffs =
+      with nixpkgs.packages.x86_64-linux;
+      with nixpkgs.builders;
+      with nixpkgs.lib;
+
+      stdenv.mkDerivation {
+        name = "dwarffs-0.1.${substring 0 8 self.lastModifiedDate}";
+
+        buildInputs = [ fuse nix nlohmann_json boost ];
+
+        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";
+
+        src = self;
+
+        installPhase =
+          ''
+            mkdir -p $out/bin $out/lib/systemd/system
+
+            cp dwarffs $out/bin/
+            ln -s dwarffs $out/bin/mount.fuse.dwarffs
+
+            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
+            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
+          '';
+      };
+
+    nixosModules.dwarffs = ...;
+
+    defaultPackage.x86_64-linux = packages.x86_64-linux.dwarffs;
+
+    checks.build = packages.x86_64-linux.dwarffs;
+  };
+}
+```
+
+A flake has the following attributes:
+
+* `description`: A short description of the flake.
+
+* `inputs`: An attrset specifying the dependencies of the flake
+  (described below).
+
+* `outputs`: A function that, given an attribute set containing the
+  outputs of each of the input flakes keyed by their identifier,
+  yields the Nix values provided by this flake. Thus, in the example
+  above, `inputs.nixpkgs` contains the result of the call to the
+  `outputs` function of the `nixpkgs` flake, and
+  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
+  attribute of `nixpkgs`.
+
+  In addition to the outputs of each input, each input in `inputs`
+  also contains some metadata about the inputs. These are:
+
+  * `outPath`: The path in the Nix store of the flake's source
+    tree. This means that you could import Nixpkgs in a more
+    legacy-ish way by writing
+
+        with import inputs.nixpkgs { system = "x86_64-linux"; };
+
+    since `nixpkgs` still contains a `/default.nix`. In this case we
+    bypass its outputs entirely and only use the flake mechanism to
+    get its source tree.
+
+  * `rev`: The commit hash of the flake's repository, if applicable.
+
+  * `revCount`: The number of ancestors of the revision `rev`. This is
+    not available for `github` repositories (see below), since they're
+    fetched as tarballs rather than as Git repositories.
+
+  * `lastModifiedDate`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
```

Comment:
Lost here as to what and why is lastModifiedDate. Is it an alternative for rev? Or is it an alternative for revCount (which then doesn't make sense for tarball/github fetcher)? Would a option tag serve the purpose?
"hopefully" and "version string" combination sounds scary, what is it expected to be used for?

Reply:
`revCount` and `lastModifiedDate` are primarily there to help you generate version strings (e.g. in the `name` attribute of a derivation). For example, the `dwarffs` package from the `dwarffs` flake has a version like `0.1.20200511`. `lastModifiedDate` is indeed an alternative to `revCount` since the latter isn't always available.

The "hopefully" refers to the fact that you can always fake timestamps with Git, but that's a case of "just don't do that" :-)